### PR TITLE
Log

### PR DESCRIPTION
**Pull request by IvanNardi**
It is quite surprising to me that a basic (and useful!) feature such as --enable-debug-messages doesn't seem to work out-of-the-box.
I don't know if I understood all the ramifications of this issue, but this PR should be an improvement to the current situation, at least

### CODE REVIEW

File: src/lib/ndpi_main.c [hunk 1]
```diff
@@ -1954,7 +1954,8 @@ struct ndpi_detection_module_struct *ndpi_init_detection_module(ndpi_init_prefs
 
   if(ndpi_str == NULL) {
 #ifdef NDPI_ENABLE_DEBUG_MESSAGES
-    NDPI_LOG_ERR(ndpi_str, "ndpi_init_detection_module initial malloc failed for ndpi_str\n");
+    /* We can't use NDPI_LOG* functions yet! */
+    fprintf(stderr, "ndpi_init_detection_module initial malloc failed for ndpi_str\n");
```
Comment (utoni) [inline src/lib/ndpi_main.c]: A shared object should never assume that file descriptor 2 is a stderr fd, it could be anything, a fd to /dev/null, a socket, pipe or whatever fd.

nDPI already has a quite nice function called `ndpi_debug_printf`, which callback can be altered with `set_ndpi_debug_function`

File: src/lib/ndpi_main.c [hunk 2]
```diff
@@ -1963,6 +1964,7 @@ struct ndpi_detection_module_struct *ndpi_init_detection_module(ndpi_init_prefs
 
 #ifdef NDPI_ENABLE_DEBUG_MESSAGES
   set_ndpi_debug_function(ndpi_str, (ndpi_debug_function_ptr) ndpi_debug_printf);
+  NDPI_BITMASK_RESET(ndpi_str->debug_bitmask);
```
Comment (utoni) [inline src/lib/ndpi_main.c:1967]: You've created the fancy function `ndpi_set_debug_bitmask` but not using it here. Why?

File: src/lib/ndpi_main.c [hunk 3]
```diff
@@ -6247,6 +6249,12 @@ u_int ndpi_get_ndpi_detection_module_size() {
   return(sizeof(struct ndpi_detection_module_struct));
 }
 
+void ndpi_set_debug_bitmask(struct ndpi_detection_module_struct *ndpi_str, NDPI_PROTOCOL_BITMASK debug_bitmask) {
+#ifdef NDPI_ENABLE_DEBUG_MESSAGES
```
Comment (utoni) [inline src/lib/ndpi_main.c:6255]: I would recommend to put the preprocessor conditional one line up as this function is only mandatory if `NDPI_ENABLE_DEBUG_MESSAGES` enabled.

File: example/reader_util.h [hunk 4]
```diff
@@ -336,4 +336,15 @@ float ndpi_flow_get_byte_count_entropy(const uint32_t byte_count[256], unsigned
 
 extern int nDPI_LogLevel;
 
+/* TODO: conditional compilation? */
```
Comment (utoni) [inline example/reader_util.h]: Would've make sense to me. Either use `NDPI_ENABLE_DEBUG_MESSAGES`, `DEBUG_TRACE` or `DEBUG_PAYLOAD`. It's up to you. I would recommend the first one as it should be set with `./configure --enable-debug-messages`.

File: example/reader_util.h [hunk 5]
```diff
@@ -336,4 +336,15 @@ float ndpi_flow_get_byte_count_entropy(const uint32_t byte_count[256], unsigned
...
+#if 1
+ #define LOG(log_level, args...)	\
```
Comment (utoni) [inline example/reader_util.h:340]: Why didn't you use `NDPI_LOG` ? It is already there.

File: src/lib/protocols/http.c [hunk 6]
```diff
@@ -589,8 +589,10 @@ static u_int16_t http_request_url_offset(struct ndpi_detection_module_struct *nd
   int i;
 
   NDPI_LOG_DBG2(ndpi_struct, "====>>>> HTTP: %c%c%c%c [len: %u]\n",
-		non_ctrl(packet->payload[0]), non_ctrl(packet->payload[1]),
```
Comment (utoni) [inline src/lib/protocols/http.c:592]: Why removing `non_ctrl` ?

**Review by utoni — CHANGES_REQUESTED** [review]
I'm actually not quite sure why you reinvented the wheel with `LOG()` and moved all those `NDPI_CURRENT_PROTO` after the `ndpi_api.h` include.

File: src/lib/ndpi_main.c (hunk 1 above)
Comment (IvanNardi) [reply src/lib/ndpi_main.c]: AFAIU, it is a little bit more complicated.
This log should be reported at the beginning of ndpi_init_detection_module(), which initializes the library and must be the first function called by the user.
At that point:
* you don't have a detection_module context, so you can't log anything via NDPI_LOG* macros
* the user can't have already called set_ndpi_debug_function(), so a custom log function isn't available yet. We could use the default implementation of ndpi_debug_printf(), but it prints on stdout. Your valid concern about stderr, apply to stdout, too.

Since this allocation error should be uncommon (to say the least), I think we have only two options:
1) don't log anything
2) directly print on stdout or stderr, having such printf protected via NDPI_ENABLE_DEBUG_MESSAGES define

I have no strong opinion one way or the other

File: src/lib/ndpi_main.c (hunk 2 above)
Comment (IvanNardi) [reply src/lib/ndpi_main.c:1967]: This is the initialization, i.e. the default value.
I think, it is better to have a proper value in ndpi_str->debug_bitmask even if the user doesn't call ndpi_set_debug_bitmask()

File: src/lib/ndpi_main.c (hunk 3 above)
Comment (IvanNardi) [reply src/lib/ndpi_main.c:6255]: It is matter of taste. I think that having an ABI that doesn't depend too much on defines is preferable from an user point-of-view.
In the same .c file, set_ndpi_debug_function() has been already implemented with the same logic (I originally copied-and-pasted from it) but I see that the library is not always consistent about that.
No strong opinion

File: example/reader_util.h (hunk 4 above)
Comment (IvanNardi) [reply example/reader_util.h]: Ok, thanks for the suggestion

File: example/reader_util.h (hunk 5 above)
Comment (IvanNardi) [reply example/reader_util.h:340]: I thought I explained it in the first commit message, but probably I didn't make myself clear. Let me rephrase it.
From 66b759ca6, the ndpi_detection_module_struct object is an opaque data type for the user so you can't dereference it in reader_utils.c code; if you try, you get a compiler error, something similar to

../src/include/ndpi_define.h:188:26: error: invalid use of undefined type ‘struct ndpi_detection_module_struct’
  188 |     if(mod != NULL && mod->ndpi_debug_printf != NULL)                           \
      |                          ^~
reader_util.c:1775:4: note: in expansion of macro ‘NDPI_LOG’
 1775 |    NDPI_LOG(0, workflow->ndpi_struct, NDPI_LOG_DEBUG, "\n\nWARNING: IPv4 fragments are not handled by this demo (nDPI supports them)\n");
      |    ^~~~~~~~

File: src/lib/protocols/http.c (hunk 6 above)
Comment (IvanNardi) [reply src/lib/protocols/http.c:592]: I didn't. The function non_ctrl() has been removed on 8b1c68ae, one year ago!
This is the perfect example to show how the debug message stuff has been broken for a long time. If --enable-debug-messages flag had worked as intended, anyone compiling ndpi with such flag in the last year would have received a compiler error.
I don't think that for such low-level message we need non_ctrl() again, but if anyone thinks otherwise, I could definitely reintroduce it

**Comment by IvanNardi** [comment]
Thanks for the review and for your time
I'll upload a new version in a few days, after having waited for other comments

File: src/lib/ndpi_main.c (hunk 1 above)
Comment (utoni) [reply src/lib/ndpi_main.c]: I see the issue. We could add an "early-logging-callback" as parameter for `ndpi_init_detection_module`.
However this would be an api break.

**Comment by lucaderi** [comment]
@lnslbrty Are we ready to merge this PR?

File: example/ndpiReader.c [hunk 7]
```diff
@@ -740,6 +740,10 @@ static void parseOptions(int argc, char **argv) {
       }
       break;
 
+    case 'u':
+      _debug_protocols = strdup(optarg);
```
Comment (utoni) [inline example/ndpiReader.c:748]: possible leak

**Review by utoni — CHANGES_REQUESTED** [review]
LGTM

File: example/ndpiReader.c (hunk 7 above)
Comment (IvanNardi) [reply example/ndpiReader.c:748]: Nice catch

**Comment by lucaderi** [comment]
Thank you all
